### PR TITLE
Automatically refresh and expire the torrent status cache

### PR DESCRIPTION
**Pull request by gazpachoking**
Stop at ratio was not working when no clients were connected, because
it was using a cached version of the torrent status, and never calling
for a refresh. When a client connected, it called for the refresh and
started working properly.

This fix turns `Torrent.status` into a property, which checks when the last refresh was, and automatically refreshes before returning if it was not recent (currently 5 seconds.) When setting the property it automatically stores the update time for the cache. Any place that wants `torrent.handle.status()` should instead use `torrent.status`, or `torrent.update_status()`, (which also returns said updated status,) after this change.

Closes: https://dev.deluge-torrent.org/ticket/3497

Questions:
- Is this method reasonable?
  _Seems like it_
- How long should it be before the status cache is considered expired?
  _Chose 5 seconds for now_
- There are a few methods in `Torrent` which grab `self.handle.status()` directly. Should those be converted to use the cache? If not, they should be changed to update the cache, to avoid extra cache refreshes.
  _Converted any of these to use `torrent.update_status()` which refreshes the cache (and returns the new status)_

### CODE REVIEW

**Comment by cas--** [comment]
Nice catch, I have been meaning to investigate that one properly

**Comment by gazpachoking** [comment]
I'm looking at all the other bits that call the live handle.status(), and having trouble deciding whether they really need a fresh status or the cached one is good now that it's guaranteed to be relatively recent. If we do want to pull guaranteed fresh status, maybe we should add something like `torrent.get_status(update=False)` That way we are guaranteed to update the cache every time `handle.status()` is called, rather than calling it sometimes and not storing it to the cache.

**Comment by gazpachoking** [comment]
I had a go at answering my own questions raised in the OP, and implemented them as such. Happy to reconsider if anyone has any better plans.

**Comment by gazpachoking** [comment]
Another thing that could be done to fix this, potentially alongside this change, is to call get_torrents_status on all of the torrents we need to check rather than looping over them one by one. Presumably the bulk call to libtorrent is more efficient that getting the status on each torrent individually. Maybe leave the current cache expiry changes in as backup, but also add the get_torrents_status call to do the bulk update for the stop ratio check?

File: deluge/core/torrentmanager.py [hunk 1]
```diff
@@ -279,19 +279,14 @@ def update(self):
                 'Paused',
                 'Queued',
             ):
-                # If the global setting is set, but the per-torrent isn't...
-                # Just skip to the next torrent.
-                # This is so that a user can turn-off the stop at ratio option on a per-torrent basis
-                if not torrent.options['stop_at_ratio']:
```
Comment (cas--) [inline deluge/core/torrentmanager.py:285]: Why was this removed?

Comment (gazpachoking) [reply deluge/core/torrentmanager.py:285]: It was never activated, because we are inside an if block that already checks it:
https://github.com/deluge-torrent/deluge/blob/b59549eaf2a222679a7cc53d8a95dbb63c0daf36/deluge/core/torrentmanager.py#L276

File: deluge/core/torrent.py [hunk 2]
```diff
@@ -685,7 +687,7 @@ def force_error_state(self, message, restart_to_resume=True):
             restart_to_resume (bool, optional): Prevent resuming clearing the error, only restarting
                 session can resume.
         """
-        status = self.handle.status()
+        status = self.update_status()
```
Comment (cas--) [inline deluge/core/torrent.py:690]: Naming of `update_status` seems misleading, `get_status` or `fetch_status` perhaps? Maybe with a `cached=False` arg?

File: deluge/core/torrent.py [hunk 3]
```diff
@@ -1024,13 +1026,35 @@ def get_status(self, keys, diff=False, update=False, all_keys=False):
...
+        """Get the torrent status fresh, not from cache.
+
+        This should be used when a guaranteed fresh status is needed rather than
+        `torrent.handle.status()` because it will update the cache as well.
+        """
+        self.status = self.handle.status()
+        return self.status
+
+    @property
+    def status(self) -> 'lt.torrent_status':
```
Comment (cas--) [inline deluge/core/torrent.py:1039]: I'm not sure if it would make it more or less complex to use `cached_property` but moving to a status property looks good :+1:

File: deluge/core/torrent.py (hunk 2 above)
Comment (gazpachoking) [reply deluge/core/torrent.py:690]: Yeah, I did `get_status` at first, until I realized there already is a `get_status` which is a combined options and status dict. Maybe rename `torrent.status` cache to `torrent.lt_status`, and have `torrent.get_lt_status`?

File: deluge/core/torrent.py (hunk 3 above)
Comment (gazpachoking) [reply deluge/core/torrent.py:1039]: Took a peek at cached_property, doesn't look like it helps us much.

File: deluge/core/torrent.py (hunk 2 above)
Comment (gazpachoking) [reply deluge/core/torrent.py]: So I did the rename to `torrent.lt_status`, with the refresh option now being `torrent.get_lt_status()`. I think this helps clarify the difference between the libtorrent object and the deluge status dict, which is grabbed by `torrent.get_status()`. Happy to do it a different way though if anyone has better ideas.

Comment (cas--) [reply deluge/core/torrent.py]: Yeah I did want to think about this some more as I am not sure that we should be exposing torrent handle status outside of the Torrent class

File: deluge/core/torrent.py [hunk 4]
```diff
@@ -234,10 +236,11 @@ def __init__(self, handle, options, state=None, filename=None, magnet=None):
         self.handle = handle
 
         self.magnet = magnet
-        self.status = self.handle.status()
+        self._lt_status: 'Optional[lt.torrent_status]' = None
+        self._lt_status_last_update: float = 0.0
 
         self.torrent_info = self.handle.get_torrent_info()
-        self.has_metadata = self.status.has_metadata
+        self.has_metadata = self.lt_status.has_metadata
```
[the remark this answers is not in the document]
Comment (gazpachoking) [reply deluge/core/torrent.py:243]: It's the `@property` I added.

File: deluge/core/torrent.py (hunk 2 above)
Comment (gazpachoking) [reply deluge/core/torrent.py]: Yeah, I was thinking that as well. It's only used in a few spots in torrentmanager, which could probably be updated.

File: deluge/core/torrent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cas--) [reply deluge/core/torrent.py:243]: doh missed that

File: deluge/core/torrent.py (hunk 2 above)
Comment (gazpachoking) [reply deluge/core/torrent.py]: The one spot in torrentmanager where it seems sorta important is here https://github.com/deluge-torrent/deluge/pull/369/files#diff-3416e08006d0ef78d79619b83bf6d42c888dc1e5b20ea03801a71e357d5484c9R1615

Which is where it handles the bulk update alert and updates the status cache of all the torrents. I think this was one of the things that improved performance on large numbers of torrents.

Comment (cas--) [reply deluge/core/torrent.py]: I am not sure about wholesale renaming here, perhaps we keep `status` attribute for now and keep new `get_lt_status`?

Comment (gazpachoking) [reply deluge/core/torrent.py]: That seems good to me too. I think perhaps `status` could/should be renamed and become private in the future, (maybe `self._lt_status`,) but I don't think that needs to be tied up in this bugfix. I'll fix the PR up when I get a chance.

Comment (gazpachoking) [reply deluge/core/torrent.py]: Reverted the `.status` rename, but kept the `.get_lt_status()` now.

**Comment by gazpachoking** [comment]
Hmm. How did I mess up the docs building?

File: deluge/core/torrent.py [hunk 5]
```diff
@@ -234,7 +236,8 @@ def __init__(self, handle, options, state=None, filename=None, magnet=None):
         self.handle = handle
 
         self.magnet = magnet
-        self.status = self.handle.status()
+        self._status: 'Optional[lt.torrent_status]' = None
```
Comment (cas--) [inline deluge/core/torrent.py:239]: Should it only be the lt.torrent_status that is the string?

**Comment by cas--** [comment]
My quick testing suggests it's related to typing added in commit [Add type hints for the torrent status methods](https://github.com/deluge-torrent/deluge/pull/369/commits/eac93ca35884ef95c37b055d12427eb0c1a40307)

I'm testing locally to see what is causing it

File: deluge/core/torrent.py (hunk 5 above)
Comment (cas--) [inline deluge/core/torrent.py:239]: Imported Optional but is string here, overzealous stringifying?

File: deluge/core/torrent.py [hunk 6]
```diff
@@ -1054,13 +1056,35 @@ def get_status(self, keys, diff=False, update=False, all_keys=False):
...
+        """Get the torrent status fresh, not from cache.
+
+        This should be used when a guaranteed fresh status is needed rather than
+        `torrent.handle.status()` because it will update the cache as well.
+        """
+        self.status = self.handle.status()
+        return self.status
+
+    @property
+    def status(self) -> 'lt.torrent_status':
```
Comment (cas--) [inline deluge/core/torrent.py:1069]: It seems that using `lt.torrent_status` here and for above method is causing the autodoc errors. Using the full `libtorrent.torrent_status` works though :shrug:

Comment (cas--) [reply deluge/core/torrent.py:1069]: Actually using `lt` module seems to be the issue here so any other string would work

Comment (cas--) [reply deluge/core/torrent.py:1069]: I added some debug to sphinx_autodoc_typehints.process_docstring and I can see the issue is that it is encountering a `Mock object` which seems to be tripping it up. This is likely due to mocking libtorrent and other 3rd party modules in sphinx config.

Comment (cas--) [reply deluge/core/torrent.py:1069]: Fixed: 62a40521789eab27341
